src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentPacket.kt

```diff
@@ -46,39 +48,24 @@ object IncomingPaymentPacket {
 
     fun decryptOnion(paymentHash: ByteVector32, packet: OnionRoutingPacket, privateKey: PrivateKey, blinding: PublicKey?): Either<FailureMessage, PaymentOnion.FinalPayload> {
         val onionDecryptionKey = blinding?.let { RouteBlinding.derivePrivateKey(privateKey, it) } ?: privateKey
-        when (val decrypted = Sphinx.peel(onionDecryptionKey, paymentHash, packet)) {
-            is Either.Left -> return Either.Left(decrypted.value)
-            is Either.Right -> {
-                if (!decrypted.value.isLastPacket) {
-                    return Either.Left(UnknownNextPeer)
-                } else {
-                    val tlvs = try {
-                        PaymentOnion.PerHopPayload.tlvSerializer.read(decrypted.value.payload.toByteArray())
-                    } catch (_: Throwable) {
-                        return Either.Left(InvalidOnionPayload(0U, 0))
-                    }
-                    val encryptedRecipientData = tlvs.get<OnionPaymentPayloadTlv.EncryptedRecipientData>()?.data
-                    val innerBlinding = tlvs.get<OnionPaymentPayloadTlv.BlindingPoint>()?.publicKey
-                    return if (encryptedRecipientData == null) {
-                        if (blinding != null || innerBlinding != null) {
-                            Either.Left(InvalidOnionBlinding(hash(packet)))
-                        } else {
-                            try {
-                                Either.Right(PaymentOnion.FinalPayload.Standard(tlvs))
-                            } catch (_: Throwable) {
-                                Either.Left(InvalidOnionPayload(0U, 0))
-                            }
+        return Sphinx.peel(onionDecryptionKey, paymentHash, packet).flatMap { decrypted ->
+            when {
+                !decrypted.isLastPacket -> Either.Left(UnknownNextPeer)
+                else -> PaymentOnion.PerHopPayload.read(decrypted.payload.toByteArray()).flatMap { tlvs ->
+                    // This blinding point is provided in the onion when we're the introduction node of a blinded path.
```

Comment:
If we are the introduction point of the blinded route, it means something went wrong. The introduction node needs to be a public node and we are not.

Reply:
Agreed, I wasn't sure whether we wanted to keep that code (mostly for testing) or not. I'll try removing it, this should also simplify that function further.

Reply:
Removed in https://github.com/ACINQ/lightning-kmp/pull/631/commits/99cf4c54514b448987bc10c5ae55f48d4fdbb6ea